Related Posts Block: make the block available even when the module is disabled

jeherve opened the pull request:
Fixes #11861

## Proposed changes:

This PR changes the behavior of the Related Posts block when the feature is not available.

### Current behavior

**On WordPress.com Simple**

On simple, you're able to disable the Related Posts feature via the `enabled` option of the `relatedposts` Jetpack option. This setting is available under Settings > Reading:

`https://yourdotcomsimplesite.com/wp-admin/options-reading.php`

When this happens, the block is still available in the editor but will not work properly, as outlined in #11861.

**On self-hosted sites**

On self-hosted sites (WoA and Jetpack), you can disable the Related Posts module, under Jetpack > Settings > Traffic

In this scenario, the block isn't registered in the editor.

### Future behavior

On both environments, the Related Posts block will remain registered and available. When the feature or the module will be disabled:

- We do not display anything on the frontend.
- In the block and site editors, we display a placeholder with a button inviting you to activate the feature. The placeholder contents and the support links should be a bit different whether you're on WordPress.com simple or self-hosted.

<img width="698" alt="Screenshot 2023-09-15 at 17 44 14" src="https://github.com/Automattic/jetpack/assets/426388/3b2562f6-2f41-44fa-9af1-02557eff4035">

- When the feature is enabled (either the module or the option), we display the block contents just like before.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* See #32401

## Does this pull request change what data or activity we track or use?

* This PR introduces new Tracks events when the placeholder is displayed, and when one enables the feature via the placeholder button.

## Testing instructions:

**This needs to be tested on simple and on self-hosted.** To test this, it's best if you use a site with a lot of existing content, so related posts can be displayed.

### On WordPress.com Simple

> **Note**
> You will need to apply this diff, as well as D121657-code to bring in the new endpoint.

1. Start a new post in the block editor, and add a 'Related Posts' block.
2. Publish and view the post to verify that the block is present on the frontend.
3. At https://wordpress.com/settings/traffic/<yourSite>, disable the Related Posts toggle
4. Reload the published posts, and verify that the Related Posts block is no longer visible
5. Edit the post again ; the block should be there, but with a placeholder content.
6. Click on the button to enable related posts back.
7. Verify that the related posts appear again on the frontend.

### On self-hosted

1. Go to Jetpack > Settings > Traffic
2. Disable the Related Posts module
3. Go to Posts > Add New
4. Add some content and save your draft.
5. Add a Related Posts Block
6. See the placeholder content
7. Enable the module from the button.
8. Publish and verify that you can see the related posts on the frontend.

jeherve committed:
Add changelog entry

jeherve committed:
Move the block registration to a file that's loaded with the block

Whether the module is active or not, we'll register the Related Posts block.

jeherve committed:
Move away from class component to function component

jeherve committed:
Display a placeholder when the module is disabled

- The placeholder includes a button inviting you to activate the module.
- While the module status loads, we display placeholder block contents.

jeherve committed:
Update test to take new RelatedPostsEdit into account

- Include module status check
- Move to function component
- Mock additional functions

jeherve committed:
Take WordPress.com into account as well

Modules behave differently there

ice9js commented on projects/plugins/jetpack/extensions/blocks/related-posts/test/edit.js:
```diff
@@ -54,29 +55,59 @@ const posts = [
+		changeStatus: jest.fn(),
+	} ),
+} ) );
+jest.mock( '@wordpress/element', () => ( {
```
Hmm, it seems mocking modules only affects the current file so this doesn't work and the tests fail. If you delete this and replace it with the following code, all is well: 

```
import { select } from '@wordpress/data';
import { store as editorStore } from '@wordpress/editor';

...

beforeEach( () => {
        ...
        
	jest.spyOn( select( editorStore ), 'getCurrentPost' ).mockReturnValue( {
		[ 'jetpack-related-posts' ]: posts
	} );
} );
```

anomiex replied:
That replacement does seem more straightforward than what's going on here.

But I think the failure is not due to anything about "only affects the current file", it's because `useSelect` comes from `@wordpress/data` rather than `@wordpress/element`. I think the wording you're looking at from https://jestjs.io/docs/jest-object#jestmockmodulename-factory-options is referring to different _test files_, not to files required by the _current_ test.

I see [extensions/blocks/gathering-tweetstorms/test/editor.js](https://github.com/Automattic/jetpack/blob/74ba77bc25f6cd5879c8996e6c82b46e0bc39c2d/projects/plugins/jetpack/extensions/blocks/gathering-tweetstorms/test/editor.js) successfully mocks `useSelect` by doing it in a different way, for example.

ice9js replied:
I tried changing `@wordpress/element` to `@wordpress/data` first but that wouldn't run at all. I guess if you mocked `useState` directly, maybe that'd work, but I didn't investigate further.

I'd argue mocking the entire `useSelect` is not a great idea as it completely bypasses any logic inside the callback itself (which in this case, there is), potentially leading to uncaught errors, more maintenance or the tests becoming unstable.

jeherve replied:
0625b7cbaab516dbdeae1ee1d9b47dac5a24fe9f did it. Thank you both!

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/related-posts-availability

jeherve committed:
Fix tests

jeherve committed:
Add new endpoints to fetch/update related posts status from option

One can also disable Related Posts (while the module is active) via a toggle in Settings > Reading in Calypso. Let's add:

- 2 new endpoints to fetch and update that info.
- a new hook to handle this new status info in the block editor.
- hook the new hook into the Related Posts' block in the editor.

jeherve committed:
Update tests to match new hook

jeherve committed:
Adjust styles

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/related-posts-availability

jeherve committed:
Ensure the link appears below the button

jeherve committed:
Merge remote-tracking branch 'origin/trunk' into update/related-posts-availability

anomiex commented:
Seems to work in self-hosted.

On Simple, when adding the block with the feature enabled I get a flash of "You'll need to activate the Related Posts feature" before it switches over to displaying the block. I think your new `useRelatedPostsStatus` needs to differentiate "loading status" from "changing status" like `useModuleStatus` does, so you can use the "loading status" indicator at https://github.com/Automattic/jetpack/blob/ad77b19b12159b7ce5bb8dc9a4d49dfd5de60864/projects/plugins/jetpack/extensions/blocks/related-posts/edit.js#L182 along with that one from `useModuleStatus`.

jeherve committed:
Consider loading state when options are being fetched

See https://github.com/Automattic/jetpack/pull/32930#issuecomment-1723922206

jeherve commented:
> On Simple, when adding the block with the feature enabled I get a flash of "You'll need to activate the Related Posts feature" before it switches over to displaying the block.

ah that's a good point, thank you! I just committed 4fda28dfd45e2b301b600ae434dd4edba2532ff5 to help with that.

anomiex reviewed (approved):
Seems to work.

sajmes commented:
On several simple sites, I am seeing the prompt to activate the related posts block in the editor, even when the setting in yoursimplesite.com/wp-admin/options-reading.php for related posts is enabled.

Clicking the activate option in the editor does nothing.

A user reported this across three sites, and I can replicate on a test site. All previously added related posts blocks are no longer working on their site.

I'm not sure if this is related to this, but it seems possible @jeherve 

Ticket: 7012332-zd-a8c

jeherve commented:
@sajmes Thanks for letting me know. Could you try on your test sites and on the customer's site again? It should work now.

sajmes commented:
@jeherve  - I checked this and it appears that for existing related post blocks to work, the block needs to be activated from any post on the site. On one of the user's sites, I activated this and the block is displaying related posts on their existing posts. 

On their other site, I did not click the activate option on the related posts block, and the site is still not displaying related posts on posts where that block was previously added.

Can you replicate that? 

Is this something we can automatically activate for existing blocks if so?

jeherve commented:
As discussed in Slack, I'm thinking that this may happen for all the folks who had added Related Posts blocks to their site without fully activating the feature in Settings > Reading.

To fix this issue, one would have to do the following:

1. Go to Settings > Reading
2. If the Related Posts are set to be displayed, set them to be hidden, and then save changes.
3. When the settings reload, mark the related posts to be displayed again.
4. Save your changes.

That should hopefully fix things.

renata-franco commented:
Is this issue related to this change?

![CleanShot 2023-09-25 at 12 47 51](https://github.com/Automattic/jetpack/assets/81378719/cdf3b72c-97b4-4288-b185-cd4fb567fe38)

jeherve commented:
@renata-franco No, this isn't related. You'll see this grayed out block whenever the block has already been inserted elsewhere on the page, and when the block has a restriction of one block per page.

renata-franco commented:
> @renata-franco No, this isn't related. You'll see this grayed out block whenever the block has already been inserted elsewhere on the page, and when the block has a restriction of one block per page.

Thank you, Jeremy! Fixed.
